remote-wallet/src/remote_wallet.rs

```diff
@@ -0,0 +1,368 @@
+        let start_time = Instant::now();
+        while start_time.elapsed() <= *max_polling_duration {
+            if let Ok(num_devices) = self.update_devices() {
+                let plural = if num_devices == 1 { "" } else { "s" };
+                trace!("{} Remote Wallet{} found", num_devices, plural);
+                return true;
+            }
+        }
+        false
+    }
+}
+
+/// `RemoteWallet` trait
+pub trait RemoteWallet {
+    /// Parse device info and get device base pubkey
+    fn read_device(
+        &self,
+        dev_info: &hidapi::HidDeviceInfo,
+    ) -> Result<RemoteWalletInfo, RemoteWalletError>;
+
+    /// Get solana pubkey from a RemoteWallet
+    fn get_pubkey(&self, derivation: DerivationPath) -> Result<Pubkey, RemoteWalletError>;
+
+    /// Sign transaction data with wallet managing pubkey at derivation path m/44'/501'/<account>'/<change>'.
+    fn sign_transaction(
+        &self,
+        derivation: DerivationPath,
+        transaction: Transaction,
+    ) -> Result<Signature, RemoteWalletError>;
+}
+
+/// `RemoteWallet` device
+#[derive(Debug)]
+pub struct Device {
+    pub(crate) path: String,
+    pub(crate) info: RemoteWalletInfo,
+    pub wallet_type: RemoteWalletType,
+}
+
+/// Remote wallet convenience enum to hold various wallet types
+#[derive(Debug)]
+pub enum RemoteWalletType {
+    Ledger(Arc<LedgerWallet>),
+}
+
+/// Remote wallet information.
+#[derive(Debug, Default, Clone)]
+pub struct RemoteWalletInfo {
+    /// RemoteWallet device model
+    pub model: String,
+    /// RemoteWallet device manufacturer
+    pub manufacturer: String,
+    /// RemoteWallet device serial number
+    pub serial: String,
+    /// Base pubkey of device at Solana derivation path
+    pub pubkey: Pubkey,
+}
+
+impl RemoteWalletInfo {
+    pub fn parse_path(mut path: String) -> Result<Self, RemoteWalletError> {
+        if is_remote_wallet_path(&path).is_ok() {
+            let path = path.split_off(6);
+            let mut parts = path.split('/');
+            let mut wallet_info = RemoteWalletInfo::default();
+            let manufacturer = parts.next().unwrap();
+            wallet_info.manufacturer = manufacturer.to_string();
+            wallet_info.model = parts.next().unwrap_or("").to_string();
+            wallet_info.pubkey = parts
+                .next()
+                .and_then(|pubkey_str| Pubkey::from_str(pubkey_str).ok())
+                .unwrap_or_default();
+            Ok(wallet_info)
+        } else {
+            Err(RemoteWalletError::InvalidPath(path))
+        }
+    }
+
+    pub fn get_pretty_path(&self) -> String {
+        format!(
+            "usb://{}/{}/{:?}",
+            self.manufacturer, self.model, self.pubkey,
+        )
+    }
+}
+
+impl PartialEq for RemoteWalletInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.manufacturer == other.manufacturer
+            && (self.model == other.model || self.model == "" || other.model == "")
```

Comment:
Ewww, this is an abuse of the PartialEq instance.  Can you `derive(PartialEq)` instead? And then, if needed, add these special cases in a utility function?